Preliminary Vulkan AS/RQ test

cmannett85-arm opened the pull request:
This test renders a small triangle that casts a shadow from a point light, onto a larger triangle.  The test runner then checks that various pixels in the final output are the correct colour.

There is also an arbitrary AS copy in the render loop just to hit more API coverage when manually capturing, but the test runner doesn't check its output.

Tested on:
* Ubuntu 24.10, RTX 3090 (ver 560.35 patch 0x300)
* Android 14, Mali-G715 (ver r49p1) - used a modified test runner but the demo code and pixel colour checks are equivalent

Core test work originally done by martyn.jacques@arm.com

baldurk commented on util/test/demos/demos.vcxproj.filters:
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Please revert whatever change this is - I think losing the BOM.

cmannett85-arm replied:
I keep seeing this on PRs but I've no idea where it's coming from, neither my Windows nor Linux machine show this change...

baldurk commented on util/test/tests/Vulkan/VK_Ray_Query.py:
```diff
@@ -0,0 +1,39 @@
+from rdtest import util as util
```
Unnecessary import here?

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+          StartUsingBackbuffer(cmd, VK_ACCESS_TRANSFER_WRITE_BIT, VK_IMAGE_LAYOUT_GENERAL);
+
+      vkCmdClearColorImage(cmd, swapimg, VK_IMAGE_LAYOUT_GENERAL,
+                           vkh::ClearColorValue(0.002f, 0.002f, 0.002f, 1.0f), 1,
```
Such a small value is not representable in an 8-bit image, and this sits close to 0.5 ULP. The test is likely passing due to ULP tolerance but this value is misleading - I don't see any reason why the background needs to be so precisely small instead of e.g. 0.1?

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+   * set by allocating from a memory group with VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT support.
+   * VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT is required for acceleration structure buffers
+   */
+  struct AddressedBuffer
```
If you update VMA you can do everything in this class using it and the existing helpers - in the demos project a significant goal is short and readable code, so using utility libraries like VMA is strongly encouraged.

If you'd like I can update it for you, to rebase the PR onto.

cmannett85-arm replied:
That would be great, thank you.  I have a vague memory of @Zorro666 doing this awhile ago but having to revert it due to a 32bit Android build issue though.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+      memoryRange.offset = 0;
+      memoryRange.size = mrq.size;
+
+      CHECK_VKR(vkFlushMappedMemoryRanges(test->device, 1, &memoryRange));
```
This code can go as in the other comment - but FYI we shouldn't handle flushing memory in the demos project - it's needed in the main RenderDoc project since we have to be super-spec careful there, but we can assume sanity in the demos project as we don't need full coverage.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+    if(!Avail.empty())
+      return;
+
+    static VkPhysicalDeviceAccelerationStructureFeaturesKHR asFeatures{
```
missing an `=` here in these declarations: `asFeatures = {`

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+
+    uint32_t indices[] = {0, 1, 2, 3, 4, 5};
+
+    uint32_t primitiveCount = 2;
```
`primitiveCount = (uint32_t)sizeof(indices) / 3` for consistency here?

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+
+    VkFormat vertexFormat = VK_FORMAT_R32G32B32_SFLOAT;
+
+    constexpr VkBufferUsageFlags blasBufferUsageFlags =
```
Naming nit: At first from the name of this I thought it was the usage flags for the buffer the BLAS was to be placed into not the inputs to the BLAS.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+
+    VkAccelerationStructureInstanceKHR asInstance = {
+        tlasTransformMatrix, 0, 0xFF, 0, VK_GEOMETRY_INSTANCE_TRIANGLE_FACING_CULL_DISABLE_BIT_KHR,
+        blasDeviceAddress};
```
For a longer in-line struct like this if you put a trailing `,` clang-format does a bit better by putting the closing `}` onto its own line.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+        vkh::BufferCreateInfo(vertexBufferSize,
+                              uniformBufferUsageFlags | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT),
+        VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    AddressedBuffer uniformIndexBuffer(
```
These are named `uniform` but they look to be normal VBs/IBs? They also don't seem to need device addresses, they could be normal buffers.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,591 @@
+     */
+    VkDescriptorPool descriptorPool = VK_NULL_HANDLE;
+
+    std::vector<VkDescriptorPoolSize> poolSizes = {{VK_DESCRIPTOR_TYPE_ACCELERATION_STRUCTURE_KHR, 1}};
```
I would prefer to add this type to the default pools conditionally with the extension - as with `VK_DESCRIPTOR_TYPE_INLINE_UNIFORM_BLOCK` descriptors.

cmannett85-arm commented on util/test/demos/CMakeLists.txt:
```diff
@@ -219,7 +220,7 @@ set(SRC main.cpp
 
 if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
```
The newer VMA requires C++17 for `aligned_alloc` support.

cmannett85-arm commented on util/test/demos/vk/vk_headers.h:
```diff
@@ -86,7 +86,7 @@
 
 // avoid warning about unused variables
-#define VMA_DEBUG_LOG_FORMAT(format, ...) (void)(__VA_ARGS__);
+#define VMA_DEBUG_LOG_FORMAT(format, ...)
```
I was getting `vk_mem_alloc.h:6607:113: error: left operand of comma operator has no effect [-Werror,-Wunused-value]` errors on Clang, so I just made this an empty macro which appeased it.

baldurk commented on util/test/demos/vk/vk_ray_query.cpp:
```diff
@@ -0,0 +1,499 @@
+    vkDestroyAccelerationStructureKHR(device, newBlas, NULL);
+    vkDestroyAccelerationStructureKHR(device, tlas, NULL);
+    vkDestroyAccelerationStructureKHR(device, blas, NULL);
+    trisIndexBuffer.free();
```
These frees are unnecessary - AllocatedBuffer's are tracked in a list and destroyed on shutdown as with other common types like pipelines, descriptor sets, etc allocated through/with the VulkanTest rather than directly.

cmannett85-arm committed:
Preliminary Vulkan AS/RQ test

This test renders a small triangle that casts a shadow from a point light, onto a larger triangle.  The test runner then checks that various pixels in the final output are the correct colour.

There is also an arbitrary AS copy in the render loop just to hit more API coverage when manually capturing, but the test runner doesn't check its output.

Core test work originally done by martyn.jacques@arm.com
